backend/src/api.ts

```diff
@@ -0,0 +1,684 @@
+  if (differenceMatch) {
+    // The difference image we calculated already existed, so we can remove this one. The
+    // `saveComparisonImage` will also generate a thumbnail already, so we have to delete
+    // that file, too
+    fs.unlinkSync(difference);
+    fs.unlinkSync(thumbnailForImage(difference));
+    // and use the old one instead
+    difference = differenceImage(group, name, hardware, differenceMatch);
+  }
+
+
+  let testData: TestData = {
+    pixelError: nPixels,
+    timeStamp: timeStamp,
+    timing: Number(timing),
+    nErrors: nLogLines,
+    commitHash: commitHash,
+    referenceImage: path.basename(reference),
+    candidateImage: candidateMatch ? candidateMatch : timeStamp,
+    differenceImage: differenceMatch ? differenceMatch : timeStamp
+  };
+
+  saveTestData(testData, testDataPath(group, name, hardware, timeStamp));
+  addTestData(group, name, hardware, testData);
+  res.status(200).end();
+}
+
+
+
+/**
+ * This API call runs a single test against the current reference image and returns the
+ * difference image back to the caller. The results of this comparison are _not_ stored on
+ * the server. The necessary test information is passed along as a JSON-encoded body, and
+ * test-related files are included as multipart-encoded files.
+ *
+ * For the body, the required fields are:
+ *   - `hardware`: The hardware on which the test was run
+ *   - `group`: The name for the test's group for which a candidate is submitted
+ *   - `name`: The name of the test for which a candidate is submitted
+ *
+ * For the files, the following are needed:
+ *   - file: The generated candidate file
+ */
+function handleRunTest(req: express.Request, res: express.Response) {
+  const hardware = req.body.hardware;
+  if (hardware == null) {
+    res.status(400).json({ error: "Missing field 'hardware'" });
+    return;
+  }
+
+  const group = req.body.group;
+  if (group == null) {
+    res.status(400).json({ error: "Missing field 'group'" });
+    return;
+  }
+
+  const name = req.body.name;
+  if (name == null) {
+    res.status(400).json({ error: "Missing field 'name'" });
+    return;
+  }
+
+  if (req.file == null) {
+    res.status(400).json({ error: "Missing field 'file'" });
+    return;
+  }
+
+  try {
+    const png = PNG.sync.read(req.file.buffer);
+    if (png.width != Config.size.width || png.height != Config.size.height) {
+      let w = Config.size.width
+      let h = Config.size.height
+      res.status(400).json({ error: `Image has the wrong size. Expected (${w}, ${h})`});
+      return;
+    }
+  }
+  catch (e: any) {
+    res.status(400).json({ error: `Error loading image: ${e}`});
+    return;
+  }
+
+  if (!hasReferenceImage(group, name, hardware)) {
+    res.status(400).json({ error: "Error comparing, no reference image found" });
+    return;
+  }
+
+  const reference = referenceImage(group, name, hardware);
+  const candidate = temporaryPath() + "/candidate.png";
```

Comment:
Here you can use string literals:

```suggestion
  const candidate = `${temporaryPath()} /candidate.png`;
```